environs/sync/sync.go

```diff
@@ -125,30 +122,45 @@ func SyncTools(syncContext *SyncContext) error {
 	}
 
 	logger.Infof("listing target agent binaries storage")
-	targetTools, err := syncContext.TargetToolsFinder.FindTools(syncContext.MajorVersion, syncContext.Stream)
-	switch err {
-	case nil, coretools.ErrNoMatches, envtools.ErrNoTools:
-	default:
-		return errors.Trace(err)
-	}
-	for _, tool := range targetTools {
-		logger.Debugf("found target agent binary: %v", tool)
+
+	result, err := sourceTools.Match(coretools.Filter{Number: syncContext.ChosenVersion})
+	logger.Tracef("syncContext.ChosenVersion %s, result %s", syncContext.ChosenVersion, result)
+	if err != nil {
+		return errors.Wrap(err, errors.NotFoundf("%q", syncContext.ChosenVersion))
+	}
+	_, chosenList := result.Newest()
+	logger.Tracef("syncContext.ChosenVersion %s, chosenList %s", syncContext.ChosenVersion, chosenList)
+	if syncContext.TargetToolsFinder != nil {
+		targetTools, err := syncContext.TargetToolsFinder.FindTools(
+			syncContext.ChosenVersion.Major, syncContext.Stream,
+		)
+		switch err {
+		case nil, coretools.ErrNoMatches, envtools.ErrNoTools:
+		default:
+			return errors.Trace(err)
+		}
+		for _, tool := range targetTools {
+			logger.Debugf("found target agent binary: %v", tool)
+		}
+		if targetTools.Exclude(chosenList).Len() != targetTools.Len() {
+			// already in target.
+			return nil
+		}
+		logger.Infof("found %d target agent binaries; %d agent binaries to be copied", len(targetTools), len(chosenList))
```

Comment:
Since we don't support "all" anymore, just a specific version, this isn't really relevant